sretoolbox/binaries/base.py

```diff
@@ -0,0 +1,243 @@
+"""
+Abstractions around system binaries.
+"""
+
+import logging
+import os
+
+from abc import ABCMeta
+from abc import abstractmethod
+from collections import Counter
+from pathlib import Path
+
+import requests
+
+from semver import VersionInfo
+
+from sretoolbox.utils import run
+
+
+LOG = logging.getLogger(__name__)
+
+
+class CommandNotFoundError(Exception):
+    """
+    Used when a binary is not available in the system.
+    """
+
+
+class Binary(metaclass=ABCMeta):
+    """
+    Represents a binary in the system. Tries to find the system binary
+    on the $PATH, in the specified version. When the binary is not
+    available locally, tryes to download it.
+
+    :param version: the version of the binary
+    """
+    binary_template = ''
+    download_url_template = ''
+
+    def __init__(self, version, download_path='/tmp'):
+        # Making sure that the version contains
+        # valid minor and patch elements
+        counter = Counter(version)
+        if counter['.'] == 0:
+            version += '.0.0'
+        elif counter['.'] == 1:
+            version += '.0'
+
+        self.expected_version = VersionInfo.parse(version=version)
+        LOG.debug('Expected %s version: %s', self.binary,
+                  self.expected_version)
+
+        self.download_path = download_path
+
+        downloaded = False
+        # First try to get the path from the system
+        self._command = self._get_command_path(self.binary)
+
+        if self._command is None:
+            # First try to download the binary
+            downloaded_file = self._download()
+            self._command = self.process_download(path=downloaded_file)
+            downloaded = True
+
+        # No luck. Binary is not on the system and it also can't
+        # be downloaded.
+        if self._command is None:
+            raise CommandNotFoundError(f'Not able to get {self.binary} '
+                                       f'version {self.expected_version}')
+
+        # Checking if we have the right version
+        result = run(cmd=self.get_version_command())
+        self._command_version = self.parse_version(version=result)
+        if not self._compare(expected=self.expected_version,
+                             actual=self._command_version):
+
+            if downloaded:
+                # If even after download, the version doesn't match,
+                # we are done trying.
+                raise CommandNotFoundError(f'Not able to get {self.binary} '
+                                           f'version {self.expected_version}')
```

Comment:
I would write something different here, to differentiate this case from the previous failure mode

Reply:
sure